src/main/java/bio/terra/service/job/JobService.java

```diff
@@ -318,89 +326,70 @@ public List<JobModel> enumerateJobs(
-            if (userRoles == null) {
-              userRoles = samService.listActions(userReq, resourceType, resourceId);
-              roleMap.put(key, userRoles);
-            }
-            if (userRoles.contains(action.toString())) {
-              flightStateList.add(flightState);
-            }
-          }
-        }
-        // Stop processing as soon as we've got enough to return
-        if (flightStateList.size() == offset + limit) {
-          break;
-        }
-      }
-      numTimesQueried++;
-      nextPageToken = filterResults.getNextPageToken();
+  private FlightBooleanOperationExpression createFlightFilter(
+      AuthenticatedUserRequest userReq, String className) {
+    List<FlightFilterPredicateInterface> topLevelBooleans = new ArrayList<>();
+    // Exclude FileIngestWorkerFlight subflights since they pollute the jobs page.  We may
+    // eventually add a boolean to re-add or display the subflights as children.
+    // Note: do not add this filter if the user is explicitly requesting FileIngestWorkerFlight's
+    if (!StringUtils.isEmpty(className)) {
+      topLevelBooleans.add(makePredicateFlightClass(FlightFilterOp.EQUAL, className));
     }
-    if (flightStateList.size() <= offset) {
-      return new ArrayList<>();
-    } else if (flightStateList.size() < offset + limit) {
-      return flightStateList.subList(offset, flightStateList.size());
+    if (!StringUtils.isEmpty(className) && !className.equals(FileIngestWorkerFlight.class.getName())
+        || StringUtils.isEmpty(className)) {
+      topLevelBooleans.add(
+          makePredicateFlightClass(FlightFilterOp.NOT_EQUAL, FileIngestWorkerFlight.class));
     }
-    return flightStateList.subList(offset, offset + limit);
+
+    // Make sure that only flights a user has access to are returned if the user is not an admin
+    if (!checkUserCanListAnyJob(userReq)) {
+      topLevelBooleans.add(
+          makeOr(
+              Stream.of(
+                      // Always allow the user to see their own flights
+                      makePredicateInput(
+                          JobMapKeys.SUBJECT_ID.getKeyName(),
+                          FlightFilterOp.EQUAL,
+                          userReq.getSubjectId()),
+                      // The user can view flights associated with profiles they have access to
+                      makeResourceTypeFilter(IamResourceType.SPEND_PROFILE, userReq),
+                      // The user can view flights associated with datasets they have access to
+                      makeResourceTypeFilter(IamResourceType.DATASET, userReq),
+                      // The user can view flights associated with snapshots they have access to
+                      makeResourceTypeFilter(IamResourceType.DATASNAPSHOT, userReq))
+                  // Remove nulls (e.g. if the user doesn't have access to any profiles or datasets)
+                  .filter(Objects::nonNull)
+                  .toArray(FlightFilterPredicateInterface[]::new)));
```

Comment:
Maybe `makeOr()` and `makeAnd()` should've had `List<T>` constructors too? Since the underlying object is `List<T>` it seems awkward to convert to `T[]` only so you convert to `List<T>` in the constructor.

Reply:
Yeah, I might go back and add that though for expedience, I'm gonna stick with arrays for now